internal/formats/common/spdxhelpers/to_syft_model.go

```diff
@@ -0,0 +1,373 @@
+	for _, r := range p.PackageExternalReferences {
+		for _, category := range categories {
+			if r.Category == category {
+				return r
+			}
+		}
+	}
+	return nil
+}
+
+func collectSyftFiles(spdxIDMap map[string]interface{}, doc *spdx.Document2_2) {
+	for _, f := range doc.UnpackagedFiles {
+		spdxIDMap[string(f.FileSPDXIdentifier)] = toSyftLocation(f)
+	}
+}
+
+func toSyftRelationships(spdxIDMap map[string]interface{}, doc *spdx.Document2_2) []artifact.Relationship {
+	var out []artifact.Relationship
+	for _, r := range doc.Relationships {
+		// FIXME what to do with r.RefA.DocumentRefID and  r.RefA.SpecialID
+		if r.RefA.DocumentRefID != "" && requireAndTrimPrefix(r.RefA.DocumentRefID, "DocumentRef-") != string(doc.CreationInfo.SPDXIdentifier) {
+			log.Debugf("relationship to external document: %+v", r)
+			continue
+		}
+		a := spdxIDMap[string(r.RefA.ElementRefID)]
+		b := spdxIDMap[string(r.RefB.ElementRefID)]
+		from, fromOk := a.(*pkg.Package)
+		toPackage, toPackageOk := b.(*pkg.Package)
+		toLocation, toLocationOk := b.(*source.Location)
+		if !fromOk || !(toPackageOk || toLocationOk) {
+			log.Debugf("unable to find valid relationship mapping from SPDX 2.2 JSON, ignoring: (from: %+v) (to: %+v)", a, b)
+			continue
+		}
+		var to artifact.Identifiable
+		var typ artifact.RelationshipType
+		switch r.Relationship {
+		case "CONTAINS":
+			if toLocationOk {
+				from.Locations = append(from.Locations, *toLocation)
+				typ = artifact.ContainsRelationship
+				to = toLocation
+			}
+		case "OVERWRITE":
+			typ = artifact.OwnershipByFileOverlapRelationship
+			to = toPackage
+		}
+		if typ != "" && to != nil {
+			out = append(out, artifact.Relationship{
+				From: from,
+				To:   to,
+				Type: typ,
+				Data: nil, // FIXME should there be anything for this Data?
+			})
+		}
+	}
+	return out
+}
+
+func collectSyftPackages(spdxIDMap map[string]interface{}, doc *spdx.Document2_2) {
+	for _, p := range doc.Packages {
+		syftPkg := toSyftPackage(p)
+		spdxIDMap[string(p.PackageSPDXIdentifier)] = syftPkg
+		for _, f := range p.Files {
+			loc := toSyftLocation(f)
+			syftPkg.Locations = append(syftPkg.Locations, *loc)
+		}
+	}
+}
+
+func toSyftLocation(f *spdx.File2_2) *source.Location {
+	return &source.Location{
+		Coordinates: source.Coordinates{
+			RealPath:     f.FileName,
+			FileSystemID: requireAndTrimPrefix(f.FileSPDXIdentifier, "layerID: "),
+		},
+		VirtualPath: f.FileName,
+	}
+}
+
+func requireAndTrimPrefix(val interface{}, prefix string) string {
+	if v, ok := val.(string); ok {
+		if i := strings.Index(v, prefix); i == 0 {
+			return strings.Replace(v, prefix, "", 1)
+		}
+	}
+	return ""
+}
+
+type pkgInfo struct {
+	purl packageurl.PackageURL
+	typ  pkg.Type
+	lang pkg.Language
+}
+
+func (p *pkgInfo) qualifierValue(name string) string {
+	for _, q := range p.purl.Qualifiers {
+		if q.Key == name {
+			return q.Value
+		}
+	}
+	return ""
+}
+
+func extractPkgInfo(p *spdx.Package2_2) pkgInfo {
```

Comment:
should this PR absorb the functionality from https://github.com/anchore/grype/pull/607 regarding extracting source information from pURLs as well? (this would need to change to account for this)

Reply:
Just so I understand -- this is what we discussed basically incorporating a [port of this commit](https://github.com/anchore/grype/pull/607/commits/7b6058049eed63907dc11b7dbdc00019bcf18e31)?

Reply:
I *think* this has been done already, please do check and tell me if I've missed something!